src/System.Linq/tests/ToArrayToListToDictionaryTests.cs

```diff
@@ -0,0 +1,469 @@
+
+        [Fact]
+        public void ToDictionary_UseDefaultComparerOnNull()
+        {
+            CustomComparer<int> comparer = null;
+            CollectionTest<int> collection = new CollectionTest<int>(new int[] { 1, 2, 3, 4, 5, 6 });
+
+            Dictionary<int, int> result1 = collection.ToDictionary(key => key, comparer);
+            Assert.Same(EqualityComparer<int>.Default, result1.Comparer);
+
+            Dictionary<int, int> result2 = collection.ToDictionary(key => key, val => val, comparer);
+            Assert.Same(EqualityComparer<int>.Default, result2.Comparer);
+        }
+
+        [Fact]
+        public void ToDictionary_KeyValueSelectorsWork()
+        {
+            CollectionTest<int> collection = new CollectionTest<int>(new int[] { 1, 2, 3, 4, 5, 6 });
+
+            Dictionary<int, int> result = collection.ToDictionary(key => key + 10, val => val + 100);
+
+            Assert.Equal(collection.Items.Select(o => o + 10), result.Keys);
+            Assert.Equal(collection.Items.Select(o => o + 100), result.Values);
+        }
+
+
+        [Fact]
+        public void ToDictionary_ThrowArgumentNullExceptionWhenSourceIsNull()
+        {
+            int[] source = null;
+            Assert.Throws<ArgumentNullException>(() => source.ToDictionary(key => key));
+        }
+
+
+        [Fact]
+        public void ToDictionary_ThrowArgumentNullExceptionWhenKeySelectorIsNull()
+        {
+            int[] source = new int[0];
+            Func<int, int> keySelector = null;
+            Assert.Throws<ArgumentNullException>(() => source.ToDictionary(keySelector));
+        }
+
+        [Fact]
+        public void ToDictionary_ThrowArgumentNullExceptionWhenValueSelectorIsNull()
+        {
+            int[] source = new int[0];
+            Func<int, int> keySelector = key => key;
+            Func<int, int> valueSelector = null;
+            Assert.Throws<ArgumentNullException>(() => source.ToDictionary(keySelector, valueSelector));
+        }
+
+
+        [Fact]
+        public void ToDictionary_KeySelectorThrowException()
+        {
+            int[] source = new int[] { 1, 2, 3 };
+            Func<int, int> keySelector = key =>
+                {
+                    if (key == 1)
+                        throw new InvalidOperationException();
+                    return key;
+                };
+
+
+            Assert.Throws<InvalidOperationException>(() => source.ToDictionary(keySelector));
+        }
+
+        [Fact]
+        public void ToDictionary_ValueSelectorThrowException()
+        {
+            int[] source = new int[] { 1, 2, 3 };
+            Func<int, int> keySelector = key => key;
+            Func<int, int> valueSelector = value =>
+            {
+                if (value == 1)
+                    throw new InvalidOperationException();
+                return value;
+            };
+
+            Assert.Throws<InvalidOperationException>(() => source.ToDictionary(keySelector, valueSelector));
+        }
```

Comment:
> But it is rather tests for Dictionary<TKey, TValue> class, not for ToDictionary method.

It's ```Dictionary<TKey, TValue>``` that currently does the argument validation, but that assumes that ```ToDictionary``` appropriately passes such values through, which of course a faulty implementation might not, so it's still good to test it via ```ToDictionary```.